proof/refine/ARM/Finalise_R.thy

```diff
@@ -3650,54 +3651,55 @@ lemma unbind_notification_corres:
-                          valid_tcb_def valid_bound_ntfn_def
-                   split: option.splits)
+          apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+         apply (wpsimp wp: gbn_wp' gbn_wp get_ntfn_ko' simp: obj_at_def split: option.split)+
+   apply (frule invs_valid_objs)
+   apply (clarsimp simp: is_tcb)
+   apply (frule_tac thread=t and y=tcb in valid_tcb_objs)
+    apply (simp add: get_tcb_rev)
+   apply (clarsimp simp: valid_tcb_def)
+   apply (metis obj_at_simps(1) valid_bound_obj_Some)
   apply (clarsimp dest!: obj_at_valid_objs' bound_tcb_at_state_refs_ofD' invs_valid_objs'
                    simp: projectKOs valid_obj'_def valid_tcb'_def valid_bound_ntfn'_def
-                         tcb_ntfn_is_bound'_def
                   split: option.splits)
-  done *)
+  done
 
 lemma unbind_maybe_notification_corres:
   "corres dc
-      (invs and ntfn_at ntfnptr) (invs' and ntfn_at' ntfnptr)
+      (valid_objs and ntfn_at ntfnptr) (valid_objs' and ntfn_at' ntfnptr)
       (unbind_maybe_notification ntfnptr)
       (unbindMaybeNotification ntfnptr)"
   apply (simp add: unbind_maybe_notification_def unbindMaybeNotification_def)
   apply (rule corres_guard_imp)
-  sorry (* unbind_maybe_notification_corres *) (*
+    apply (clarsimp simp: maybeM_def get_sk_obj_ref_def)
     apply (rule corres_split[OF _ get_ntfn_corres])
       apply (rule corres_option_split)
-        apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+        apply (simp add: ntfn_relation_def)
        apply (rule corres_return_trivial)
-      apply simp
-      apply (rule corres_split[OF _ set_ntfn_corres])
-         apply (rule sbn_corres)
-        apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
-       apply (wp get_simple_ko_wp getNotification_wp)+
-   apply (clarsimp elim!: obj_at_valid_objsE
-                   dest!: bound_tcb_at_state_refs_ofD invs_valid_objs
-                    simp: valid_obj_def is_tcb tcb_ntfn_is_bound_def
-                          valid_tcb_def valid_bound_ntfn_def valid_ntfn_def
-                   split: option.splits)
-  apply (clarsimp dest!: obj_at_valid_objs' bound_tcb_at_state_refs_ofD' invs_valid_objs'
-                   simp: projectKOs valid_obj'_def valid_tcb'_def valid_bound_ntfn'_def
-                         tcb_ntfn_is_bound'_def valid_ntfn'_def
-                  split: option.splits)
-  done *)
+      apply (simp add: bind_assoc)
+      apply (rule corres_split[OF sbn_corres])
+        apply (simp add: update_sk_obj_ref_def)
+        apply (rule_tac P="ko_at (Notification sc) ntfnptr" in corres_symb_exec_l)
+           apply (rename_tac ntfnA ntfnH tcbPtr ntfnA')
+           apply (rule_tac F="ntfnA = ntfnA'" in corres_gen_asm)
+           apply (rule set_ntfn_corres)
+           apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+          apply (wpsimp wp: simp: split: option.splits)
```

Comment:
Nitpick: the `wp: ` and `simp: ` aren' needed are they?

Reply:
In fact I don't need this line at all since there's a `wpsimp ... +` in the following line.